### PR TITLE
Refactor API / Wilcox test

### PR DESCRIPTION
**Pull request by grst**
- [x] Add minimal test dataset
- [x] Reorganize code 
- [x] new base class for generic tests, linear model base is subclass of it
- [x] new interface for simple tests
- [x] (draft) implementation of wilcoxon test
- [ ] Add tests

### CODE REVIEW

File: src/multi_condition_comparisions/methods/_base.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+            Subset anndata by a boolean mask stored in this column in `.obs` before making any tests
+        layer
+            Use this layer instead of `.X`.
+
+        Returns
+        -------
+        Pandas dataframe with results ordered by significance. If multiple comparisons were performed this
+        is indicated in an additional column.
+        """
+        ...
```
Comment (grst) [inline src/multi_condition_comparisions/methods/_base.py:116]: This is the new "simple" interface that I envisage can be supported by all methods. It is constrained in what it can do (just comparisons between groups on a single column, no complex designs, no covariates). 

It's implemented as a static method on every methods class, so you can do
```python
WilcoxonTest.compare_groups(...)
# or
PyDESeq2.compare_groups(...)
```

This kind of replaces the wrapper function that was there previously.

File: src/multi_condition_comparisions/methods/_base.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+
+
+class LinearModelBase(MethodBase):
```
Comment (grst) [inline src/multi_condition_comparisions/methods/_base.py:119]: The base class we had previously is now `LinearModelBase` that contains additional tooling tailored for methods that support a linear model interface (PyDESEq, edgeR, statsmodels, ...)

File: src/multi_condition_comparisions/methods/_edger.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+import pandas as pd
+from scanpy import logging
+from scipy.sparse import issparse
+
+from multi_condition_comparisions._util import check_is_integer_matrix
+
+from ._base import LinearModelBase
+
+
+class EdgeR(LinearModelBase):
```
Comment (grst) [inline src/multi_condition_comparisions/methods/_edger.py:11]: I split up the methods into a separate file each, the `de.py` was getting too long for my taste.

File: src/multi_condition_comparisions/methods/_simple_tests.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        x1 = self.data[comparison_idx, :]
+        res = []
+        for var in tqdm(self.adata.var_names):
+            tmp_x0 = x0[:, self.adata.var_names == var]
+            tmp_x0 = np.asarray(x0.todense()).flatten() if issparse(x0) else x0.flatten()
+            tmp_x1 = x1[:, self.adata.var_names == var]
+            tmp_x1 = np.asarray(x1.todense()).flatten() if issparse(x1) else x1.flatten()
+            pval = test_fun(x=tmp_x0, y=tmp_x1).pvalue
+            res.append({"variable": var, "pvalue": pval, "fold_change": "TODO"})
+        return pd.DataFrame(res).sort_values("pvalue").set_index("variable")
```
Comment (grst) [inline src/multi_condition_comparisions/methods/_simple_tests.py:99]: Here's a suggested implementation for the WilcoxonTest. This is completely untested as I was running out of time and `paired` tests are not supported yet, but I'm sketching how I think this could be possible.

File: tests/conftest.py [hunk 5]
```diff
@@ -28,6 +30,36 @@ def test_adata(test_counts, test_metadata):
...
+            # [gene1, gene2]
+            [10, 200],
+            [20, 220],
+            [12, 400],
+            [25, 420],
+            [13, 600],
+            [24, 620],
+        ]
+    )
+    return ad.AnnData(X=X, obs=obs, var=var)
```
Comment (grst) [inline tests/conftest.py:60]: A very minimal test dataset, it's just not used anywhere yet.

**Review by grst — COMMENTED** [review]
@Zethson, @ilan-gold Here are my suggestions on how to restructure the API. LMK what you think! 

I don't know how much I can contribute implementation-wise from next week on (starting to work regularly again), but I'll always be around for discussion and code review.

File: src/multi_condition_comparisions/methods/_base.py (hunk 1 above)
Comment (ilan-gold) [reply src/multi_condition_comparisions/methods/_base.py:116]: What is the goal of this being a classmethod?  Don't we need to be able to instantiate the `cls` with a design matrix?

File: tests/conftest.py [hunk 6]
```diff
@@ -29,5 +29,30 @@ def test_adata(test_counts, test_metadata):
...
+    obs = pd.DataFrame(
+        [
+            ["A", "D1", "X", 0.2],
+            ["B", "D1", "X", 0.7],
+            ["A", "D2", "Y", 1.6],
+            ["B", "D2", "Y", 42],
+            ["A", "D3", "Y", 212],
+            ["B", "D3", "Y", 6023],
+        ],
+        columns=["condition", "donor", "other", "cont"],
```
Comment (Zethson) [inline tests/conftest.py:42]:
```suggestion
        columns=["condition", "donor", "other", "continuous"],
```

I have to admit, it took me too many second to realize the purpose of this column

File: src/multi_condition_comparisions/methods/_base.py (hunk 1 above)
Comment (ilan-gold) [reply src/multi_condition_comparisions/methods/_base.py:116]: Ok I kind of see now what you're getting at.  No more specific function for wrapping, just static methods that don't need `design_matrix` anyway because it should be inferrable from the conditions....

Comment (ilan-gold) [reply src/multi_condition_comparisions/methods/_base.py:116]: Is that right?

Comment (grst) [reply src/multi_condition_comparisions/methods/_base.py:116]: exactly. For more complex designs, users should just use the object-oriented linear-model interface.

**Comment by ilan-gold** [comment]
@Zethson Are you saying you want to keep `run_de` and just use that instead of `compare_groups` as the "public facing API?"  I would lean more towards the class-based implementation as the main public-facing API - after all, that's what scanpy is moving towards as well.

**Comment by Zethson** [comment]
This is also what I was trying to say above @ilan-gold . I also considered only offering the classes as the single API option which would make our life a bit easier and it might be a good idea to only have one way to do things. I'm like 50/50 but if you lean more towards that, I'd totally support it

**Comment by ilan-gold** [comment]
I removed the wrapper and added a "correctness" test where we draw from two very different negative binomial distributions and then test them for different groups.  This seems to be failing for reasons I don't understand...maybe a bug but the fact that only `PyDESeq2` seems to be failing indicates this might not be the case...

File: src/multi_condition_comparisions/methods/_simple_tests.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            comparison_idx = np.where(obs_df[column] == group_to_compare)[0]
+            if baseline is None:
+                baseline_idx = np.where(obs_df[column] != group_to_compare)[0]
+            else:
+                baseline_idx = np.where(obs_df[column] == baseline)[0]
+            res_dfs.append(
+                model._compare_single_group(baseline_idx, comparison_idx).assign(
+                    comparison=f"{group_to_compare}_vs_{baseline if baseline is not None else 'rest'}"
+                )
+            )
```
Comment (grst) [inline src/multi_condition_comparisions/methods/_simple_tests.py:76]: @ilan-gold, I'm not entirly sure this works for the paired test. The indices generated here are valid for the sorted df, but the `_compare_single_group` function only has the unsorted AnnData -- so I don't think the indices are referring to the correct observations. Or am I missing something?

File: src/multi_condition_comparisions/methods/_base.py [hunk 8]
```diff
@@ -0,0 +1,333 @@
...
+            return self.adata.layer[self.layer]
+
+    @classmethod
+    @abstractmethod
+    def compare_groups(
+        cls,
+        adata: AnnData,
+        column: str,
+        baseline: str,
+        groups_to_compare: str | Sequence[str],
```
Comment (grst) [inline src/multi_condition_comparisions/methods/_base.py:80]: @ilan-gold, why did you remove the `None` option? Don't you think this could work as specified in the docstring? 
```
baseline
     baseline value (one category from variable). If set to "None" this refers to "all other categories".
 groups_to_compare
    One or multiple categories from variable to compare against baseline. Setting this to None refers to
    "all categories"
 ```

**Comment by ilan-gold** [comment]
Oh well, apparently there are other issues

File: src/multi_condition_comparisions/methods/_base.py [hunk 9]
```diff
@@ -157,17 +158,39 @@ def compare_groups(
...
+        test_kwargs: dict = None,
     ) -> pd.DataFrame:
-        raise NotImplementedError(
-            "TODO: should be possible to just initialize a model, and build a contrast. `cls` provides access to the current subclass"
+        if test_kwargs is None:
+            test_kwargs = {}
+        if fit_kwargs is None:
+            fit_kwargs = {}
+        if paired_by is not None:
+            warnings.warn("Cannot use `paired_by` with linear tests.  Ignoring paramere", UserWarning, stacklevel=2)
```
Comment (grst) [inline src/multi_condition_comparisions/methods/_base.py:175]: You can, just use `f"~{column} + {paired_by}"` as the formula.

File: src/multi_condition_comparisions/methods/_base.py (hunk 8 above)
Comment (ilan-gold) [reply src/multi_condition_comparisions/methods/_base.py:80]: The `None` argument would involve a bunch of `DataFrame` wrangling that seems out of scope for getting this thing done.  We can always iterate, but having to handle this seemed out of scope.

File: src/multi_condition_comparisions/methods/_base.py (hunk 9 above)
Comment (ilan-gold) [reply src/multi_condition_comparisions/methods/_base.py:175]: Do we want to allow this?  Wouldn't this be testing groups of 2 against each other?

File: src/multi_condition_comparisions/methods/_base.py (hunk 8 above)
Comment (ilan-gold) [reply src/multi_condition_comparisions/methods/_base.py:80]: Maybe there's a different way to do this, but it seems like we would need to form a new column based on the polymorphic `groups_to_compare` argument, which could also potentially be None.  So there seemed to be a lot of different guard clauses we would need and behavior to deal with.

Comment (grst) [reply src/multi_condition_comparisions/methods/_base.py:80]: fair enough

**Comment by ilan-gold** [comment]
@Zethson @grst Outstanding issues:

1. Paired EdgeR testing is not working.  Here's the commit with the fixture: https://github.com/scverse/multi-condition-comparisions/pull/33/commits/bc6d8aefd134dddffa23e64307ab8ac3c222c667.  What's failing is a false positive (i.e., low p-value) result of a gene whose expression across the two groups, A and B, comes from identical mixtures of two negative binomial distributions.   These are then paired in order down the dataframe.  
2. Both pydeseq2 methods fail, and the resulting dataframe looks completely wrong (reversed i.e., gene that should have lower p-value has higher).  I have no idea why this is happening.  I have looked at this a bunch and the fact that the unpaired test is not even passing (unlike all other methods) makes me think something else is going on here.

Maybe a paired coding session to walk through it tomorrow morning @Zethson?

**Comment by grst** [comment]
maybe even just merge this and follow up on the fixes in separate PRs? Easier to review smaller units of change and this PR was mostly about the restructuring. If PyDESeq2 is broken, it's probably not the fault of this PR.

**Comment by ilan-gold** [comment]
@grst 100% agreed.  I do think that whatever is broken here is probably not a bug from us (although I have been known to be wrong in the past!).  I've already reported to the PyDESeq2 people
